primitives/src/merkle_tree/field_based_mht/mod.rs

```diff
@@ -1,347 +1,160 @@
-use crate::{
-    crh::FieldBasedHash, Error
-};
-use super::*;
+pub mod path;
+pub use self::path::*;
+
+pub mod naive;
+pub use self::naive::*;
+
+pub mod optimized;
+pub use self::optimized::*;
 
 pub mod parameters;
 pub use self::parameters::*;
 
-pub trait FieldBasedMerkleTreeConfig {
-    const HEIGHT: usize;
-    type H: FieldBasedHash;
+use algebra::{
+    Field, ToBytes, FromBytes,
+};
+use crate::{FieldBasedHash, BatchFieldBasedHash, Error, FieldBasedHashParameters, MerkleTreeError};
+use std::{clone::Clone, fmt::Debug};
+
+/// Definition of parameters needed to implement and optimize a Merkle Tree whose nodes and leaves
+/// are Field elements. The trait is generic with respect to the arity of the Merkle Tree.
+pub trait FieldBasedMerkleTreeParameters: 'static + Clone {
+    type Data: Field; /// Actually unnecessary, but simplifies the overall design
+    type H: FieldBasedHash<Data = Self::Data>;
```

Comment:
In future, the "linear" FieldBasedHash will not be the right one.
We will have a node hash function which doesn't serve the update()/finalize() functionalities.